rfcs/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.md

```diff
@@ -0,0 +1,166 @@
+
+This document describes how we will implement a mechanism for updating the
+staging database with the latest data from the production database.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+The final product of this plan will be a DAG (scheduled for `@monthly`) which
+will recreate the staging database from the most recent snapshot of the
+production API database. This will be accomplished by:
+
+1. Determine the most recent automated **production** snapshot using boto3's
+   [`describe_db_snapshots`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds/client/describe_db_snapshots.html)
+   function (an Airflow operator does not exist for this operation). This step
+   should also check the `Status` value of the response for the most recent
+   snapshot to ensure that it is `available`.
+2. In the case where the snapshot is not yet available, wait for the status of
+   the most recent snapshot using the
+   [`RdsSnapshotExistenceSensor`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/sensors/rds/index.html#airflow.providers.amazon.aws.sensors.rds.RdsSnapshotExistenceSensor).
+3. Create a new database from this snapshot using boto3's
+   [`restore_db_instance_from_db_snapshot`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds/client/restore_db_instance_from_db_snapshot.html)
+   function (an operator does not exist for this operation). This database will
+   be named in a way that does not conflict with the existing dev database name,
+   e.g. `dev-next-openverse-db`. Careful consideration will need to be made when
+   drafting this step to ensure that the generated database matches the settings
+   for the existing database exactly. Specifically, the following should be
+   matched:
+   - Availability should be a single DB instance (not multi-AZ)
+   - It should be attached to our default VPC
+   - It should not be publicly accessible
+   - It will need access to the following VPC security groups: `default`,
+     `openverse-vpc-db-access`, `staging-dev-rds-sg`
+   - It should match the dev database's instance size (currently `m5.xlarge`)
+   - It should have 3000 GB of allocated gp2 storage
+   - It should use password authentication (this may not need to be changed from
+     the default, as the snapshot should contain the same connection
+     information)
+4. Wait for the new database to be ready using the
+   [`RdsDbSensor`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/sensors/rds/index.html#airflow.providers.amazon.aws.sensors.rds.RdsDbSensor)
+   ([example](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/operators/rds.html#howto-sensor-rdsdbsensor)).
+5. Rename the old staging database to `dev-old-openverse-db` using boto3's
```

Comment:
Probably worth noting that we need to pass `ApplyImmediately` to the `modify_db_instance` calls, otherwise RDS will wait until the next maintenance window to make the name change.

It occurred to me that maybe we can just use the maintenance window to avoid some downtime. If we queue the rename of the old database, then queue the rename of the new one, RDS would handle the swap over for us and probably do so faster. However, deleting a db instance always happens immediately, so we can't queue it for the maintenance window _and_ even if we could, doing this would incur the cost of two RDS db instances until the maintenance window rather than just for the brief period of overlap time in this plan.

Anyway, just wanted to note the need for ApplyImmediately as it's not necessarily obvious reading the documentation and would be easy to miss.

Reply:
Oh excellent call, I'll be sure to add this. That's good to be aware of, but yes in this case I'd rather just incur the downtime.